cpp/src/reductions/reductions.cpp

```diff
@@ -98,11 +97,7 @@ std::unique_ptr<scalar> reduce_aggregate_impl(
     }
     case aggregation::NUNIQUE: {
       auto nunique_agg = static_cast<cudf::detail::nunique_aggregation const&>(agg);
-      return cudf::make_fixed_width_scalar(
-        cudf::detail::distinct_count(
-          col, nunique_agg._null_handling, nan_policy::NAN_IS_VALID, stream),
-        stream,
-        mr);
+      return nunique(col, nunique_agg._null_handling, output_dtype, stream, mr);
```

Comment:
Maybe if `output_dtype.id() == type_id::EMPTY` we default to `type_id.INT32` and make this PR non-breaking?

Reply:
The output-type is not optional so this suggestion would not keep this from breaking.

Reply:
No. I don't think optional is correct here.

Reply:
Thanks for clarifying. Would you mind elaborating a bit on why you think making it optional isn't appropriate here? Needless to say, this discussion is nonblocking 😄

Reply:
It is a problem that we have a single reduce API for all aggregations. Other libraries usually have separate APIs for each (or most). Ironically, the implementation under `cudf::reduce` calls individual `cudf::detail` APIs for each aggregation so the main reduce is merely a gateway. This means the parameters for reduce have to serve all aggregations and in some cases `output_type` is required and others it is ignored. So making it optional for all aggregations would not be correct. And I refuse to use EMPTY has some other type of optional indicator.
There are 2 `cudf::reduce` APIs where one has an optional initial-value. This is a good use of optional since this parameter is optional for all aggregations that can use it and it is ignored for those that do not.

Reply:
I see, thanks for explaining that